### PR TITLE
Code generator experimentation

### PR DESCRIPTION
**Pull request by katis**
I also happened to experiment with source_gen and code generation (just noticed #45).

Very quick&dirty implementations for `@observable` and `@computed`. I used Atoms instead of
Observables for fields, as the field get's created in the superclass anyways, and using it makes the initialization simple. Also, listening (with `observe(listener)`) the Observable is not possible because it's field would be hidden anyway.

The `mobx_codegen` project contains the `source_gen` generator.

The `example` project has an example of configuring it to a project and running `pub run build_runner build` will re-run code generation.

The generator creates an inherited class that overrides annotated fields/methods and makes them reactive:

```dart
// generator_example.dart
import 'package:mobx/mobx.dart';

part 'generator_example.g.dart';

@observable
abstract class User {
  User._();
  factory User() = _$User;

  @observable
  String firstName = 'Jane';

  @observable
  String lastName = 'Doe';

  @computed
  String get fullName => '$firstName $lastName';
}
```

=>

```dart
// GENERATED CODE - DO NOT MODIFY BY HAND

part of 'generator_example.dart';

// **************************************************************************
// ObservableGenerator
// **************************************************************************

class _$User extends User {
  _$User() : super._() {
    _$fullNameComputed = Computed(() => super.fullName);
  }

  final _$firstNameAtom = Atom(name: 'User.firstName');

  @override
  String get firstName {
    _$firstNameAtom.reportObserved();
    return super.firstName;
  }

  @override
  set firstName(String value) {
    super.firstName = value;
    _$firstNameAtom.reportChanged();
  }

  final _$lastNameAtom = Atom(name: 'User.lastName');

  @override
  String get lastName {
    _$lastNameAtom.reportObserved();
    return super.lastName;
  }

  @override
  set lastName(String value) {
    super.lastName = value;
    _$lastNameAtom.reportChanged();
  }

  Computed<String> _$fullNameComputed;

  @override
  String get fullName => _$fullNameComputed.value;
}
```

### CODE REVIEW

**Comment by copydog** [comment]
I noticed you create an `Atom` for each user-defined field. (I know this is an experiment)
On future release, we should create an `Observable` for each field, because `Observable` can check new value automatically. Am I correct?

**Comment by katis** [comment]
> 
> 
> I noticed you create an `Atom` for each user-defined field. (I know this is an experiment)
> On future release, we should create an `Observable` for each field, because `Observable` can check new value automatically. Am I correct?

I think that `Atom` might be the better way to go for `@observable`. Writing `@observable String foo = 'foo';` in the store class means that a field with storage is created for the `foo` variable even if you override it in the generated class. If we then create a hidden field `Observable<T>` and store the value inside it, the `super.foo` field goes to waste. It might then need to be nulled to clear the memory of the initial value, and the `Observable` value should be initialized from it's value, making the code a bit more complex.

`Observable` is pretty much an `Atom` + field for the value + event listeners. Since the class already has a field, and we can't use event listeners if the `Observable` value is hidden, using an Atom makes sense.

**Comment by pavanpodila** [comment]
I think there should also be a provision to get the underlying atom for observation, say `$field`. This will be needed if someone wanted to attach `onObserve`, `onUnobserve` handlers.

**Comment by copydog** [comment]
@pavanpodila Good point, I spend half day to think over it. I think we should keep generated `Observable` or `Atom` public-accessible by using `$field` instead of `_field`.

In order to make it self-accessible, we should generate a parent class not a sub-class (a mixin can not have a constructor that we need), then use it like `User extends _$User`

**Comment by katis** [comment]
@pavanpodila this PR is a pretty good starting point for the code generator, it has a separate package for it, build runner configuration and an example that uses code generation. Could we just merge this and then we'll have a base to build on?

**Comment by pavanpodila** [comment]
There is a drop in code coverage. Will to fix that before I can merge

**Comment by katis** [comment]
> 
> 
> Looking great🥇. I would love to see `@action` implemented as well in the next PR 🚀

I was actually working on `@action`, it's now available in this branch, but the `ActionController`, `start/end` model needs to be discussed.

I'll get to work on some tests, I don't have any idea how to test code generators though. Might have a look how `built_value` does it's tests.

File: mobx/lib/src/api/action.dart [hunk 1]
```diff
@@ -33,7 +33,7 @@ import 'package:mobx/src/core.dart';
 /// notifications and propagates them only after the completion of the action. Actions
 /// can also be nested inside, in which case the change notification will propagate when
 /// the top-level action completes.
-Action action(Function fn, {String name, ReactiveContext context}) =>
+Action createAction(Function fn, {String name, ReactiveContext context}) =>
```
Comment (pavanpodila) [inline mobx/lib/src/api/action.dart]: Why not just expose `Action()` directly and handle the null context inside? That way we don't need a `createAction`

Comment (katis) [reply mobx/lib/src/api/action.dart]: Added the factory for Action

**Comment by katis** [comment]
@pavanpodila can we merge this now?
The generator will change a lot, and the only other coverage change seems to be kind of false positive from the private constructors of annotations.

File: mobx/README.md [hunk 2]
```diff
@@ -13,56 +12,52 @@
...
 
-
 ## Cross cutting features
 
-- [x] Observability API for `ObservableValue` and `ComputedValue` 
-    - [x] `observe`
-    - [x] `intercept`
-    - [x] `onBecomeObserved`
-    - [x] `onBecomeUnobserved`
+- [x] Observability API for `ObservableValue` and `ComputedValue`
```
Comment (pavanpodila) [inline mobx/README.md]: Have to rename here as well `ObservableValue` => `Observable`

Comment (katis) [reply mobx/README.md]: Renamed

**Comment by katis** [comment]
@pavanpodila renames are done

**Comment by pavanpodila** [comment]
@katis there is a drop in test coverage but we can address that in a separate PR. I had set the branch rules to require not to have a drop in coverage. I have now relaxed that rule. Merging for now.

**Pull request merged**